### PR TITLE
Websockets support based on unicorn-binance-websocket-api

### LINKED ISSUE #118: Switch to Websocket API
To allow a higher api request volume, we should use Binance's websocket APIs instead of the current implementation.

### PR DESCRIPTION
**Pull request by idkravitz**
This PR is yet another take on websockets, directly based on @erenatas PR, which itself is based on @j-waters original implementation. The reason for this implementation to exist is an attempt to solve the websockets related issue with "silent stucks", when desync happens between user state on server and on bot.

Core changes:
- Switched from python-binance websockets backend to [unicorn-binance-websocket-api](https://github.com/oliver-zehentleitner/unicorn-binance-websocket-api/) backend. As it supposed to be more popular, polished and reliable. This switch didn't solved the issue in question, but it provided new useful logs that allowed to find the causes of desync.
- Fixed desync of user balances. Since first @j-waters implementation the bot operates on one more working thread, that manages the filling of caches with data from websocket streams. There was an issue with writes from two threads to same shared dict, which represents cache of wallet balances. In this version this accesses are implemented with mutex syncing.
- Fixed desync of orders progress. No matter what backend to use, there is a possibility to be disconnected from server and reconnect a bit later. In between disconnect and future connect server may miss a push of state updates to bot, which won't be resent on bot reconnect. This possibility is now eliminated by listen for connect events and refetching data with classic REST API

As a result of these core changes I suppose that issue to be solved.

The reasons to stay on unicorn-binance-websocket-api:
- Prompt logging, we can see how our network connections currently behave, how often they are forced to reconnect, etc.
- Forced sequencing of incoming data and back-pressure buffer. These two make our sync handling on reconnect so much easier than it could be on python-binance.
- Vast documentation, examples and community support

Minor changes:
- Apprise upgrade to 0.9.2. The author of apprise fixed issue with telegram escaping of single quotes. He introduced another issue with treating \<any text\> as html tags, but I commited a workaround for coins representation, they are now surrounded with brackets []. This fix can be reverted in 0.9.3, since author already fixed that issue in master :)
- Overall upgrade of dependencies versions, week of testing showed no issues with it.
- Fixed printing of scout message in multiple_coins_strategy, I took it from pinned message from #help-bugs-support channel since this fix is so popular :)

### CODE REVIEW

File: binance_trade_bot/binance_stream_manager.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        self.pending_orders.remove(self.tag)
+
+
+class BinanceStreamManager:
+    def __init__(
+        self, cache: BinanceCache, api_key: str, api_secret: str, binance_client: binance.client.Client, logger: Logger
+    ):
+        self.cache = cache
+        self.logger = logger
+        self.bwApiManager = BinanceWebSocketApiManager(output_default="UnicornFy", enable_stream_signal_buffer=True)
```
Comment (j-waters) [inline binance_trade_bot/binance_stream_manager.py]: Attributes should use underscores instead of camel case

**Review by j-waters — APPROVED** [review]
I'll try giving this a test soon, but from my initial look through it looks good. I think properly dealing with the multithreaded nature of the application is really important, which it looks like you've done!

File: binance_trade_bot/binance_stream_manager.py (hunk 1 above)
Comment (idkravitz) [reply binance_trade_bot/binance_stream_manager.py]: Didn't noticed it, pushed attribute naming fix

**Comment by idkravitz** [comment]
Rebased once again and fixed lint error with amending

**Review by rel-s — APPROVED** [review]

**Review by j-waters — APPROVED** [review]

**Comment by ionsome** [comment]
I wouldn't recommend you to use unicorn-binance-websocket-api. It's highly non-optimized and absolutely non-pythonic.
Despite it can solve some problems, It can become a reason for new ones.

**Comment by idkravitz** [comment]
> I wouldn't recommend you to use unicorn-binance-websocket-api. It's highly non-optimized and absolutely non-pythonic.
> Despite it can solve some problems, It can become a reason for new ones.

Can you elaborate your point? While being non-pythonic is mostly the matter of taste, the efficiency and optimization is important. What efficiency issues you've met or if you saw benchmarks or issues where ubwa is loser, please post a link here.

**Comment by ionsome** [comment]
> > I wouldn't recommend you to use unicorn-binance-websocket-api. It's highly non-optimized and absolutely non-pythonic.
> > Despite it can solve some problems, It can become a reason for new ones.
> 
> Can you elaborate your point? While being non-pythonic is mostly the matter of taste, the efficiency and optimization is important. What efficiency issues you've met or if you saw benchmarks or issues where ubwa is loser, please post a link here.

I have tested ubwa on VM with CPU limits and I can't say I was satisfied. 
It has a weird asynchronous architecture with lots of running Threads and Lock. In addition it spawns its own eventloop :D
As a result, it could be a problem of performance if you wanted to switch to asyncio.

And I think currently ubwa is hard to maintain because codebase is awful. I don't think it's good to rely on it.

Example: oliver-zehentleitner/unicorn-binance-websocket-api#162

**Comment by idkravitz** [comment]
> 
> 
> > > I wouldn't recommend you to use unicorn-binance-websocket-api. It's highly non-optimized and absolutely non-pythonic.
> > > Despite it can solve some problems, It can become a reason for new ones.
> > 
> > 
> > Can you elaborate your point? While being non-pythonic is mostly the matter of taste, the efficiency and optimization is important. What efficiency issues you've met or if you saw benchmarks or issues where ubwa is loser, please post a link here.
> 
> I have tested ubwa on VM with CPU limits and I can't say I was satisfied.
> It has a weird asynchronous architecture with lots of running Threads and Lock. In addition it spawns its own eventloop :D
> As a result, it could be a problem of performance if you wanted to switch to asyncio.
> 
> And I think currently ubwa is hard to maintain because codebase is awful. I don't think it's good to rely on it.
> 
> Example: [oliver-zehentleitner/unicorn-binance-websocket-api#162](https://github.com/oliver-zehentleitner/unicorn-binance-websocket-api/issues/162)

I replied there that since after Exception happened, message is pushed pack to process it again in next iteration, it may be a simple infinite loop in user code. Unfortunately that guy didn't provided any more info.

As for python-binance. Well we have this https://github.com/oliver-zehentleitner/unicorn-binance-websocket-api/issues/42 in ubwa and in python-binance it is mentioned NOWHERE that user MUST handle reconnects as data may be lost. Well, they don't even have a way to notify user about reconnection happened anyway. The only way to do that is to inherit from their classes and override protected methods, which aren't guaranteed to not change from version to version, etc. So current userData websockets in python-binance are broken. They were broken back then and they are sill broken on `1.0.10` version.

**Comment by j-waters** [comment]
I did experiment with ubwa when I was working on this. I agree with your points @Ma007ks, it really really isn't an ideal library but @idkravitz is right in that python-binance would require modifications to work properly. [I did make some modifications](https://github.com/edeng23/binance-trade-bot/pull/328/files/97b82f331e046f8e76d96359a53188e4b1df92cd#diff-4c09765099b368d1c4a2e5a45ac64246a52ba36ce0ddcd3c1079ad3e2ff282d3R12) but that is never a good solution and I didn't make enough to fix the problems completelty.

As I see it there are 5 options:

1. Switch back to python-binance, extending it's classes to fix the connection problems
2. Switch to our own forked version of python-binance, which fixes the problems and would one day hopefully be merged into the base library
3. Do number 2 but wait for the official python-binance library to accept our changes before we merge this PR
4. Build our own binance websocket implementation, either in a separate repo or within this one
5. Just merge this PR

I would personally say we do number 5, since this is a major feature that the community is waiting for, and there are a lot of changes that we don't want to have to keep rebasing. Down the road hopefully we can do one of the other options.

**Comment by erenatas** [comment]
> I did experiment with ubwa when I was working on this. I agree with your points @Ma007ks, it really really isn't an ideal library but @idkravitz is right in that python-binance would require modifications to work properly. [I did make some modifications](https://github.com/edeng23/binance-trade-bot/pull/328/files/97b82f331e046f8e76d96359a53188e4b1df92cd#diff-4c09765099b368d1c4a2e5a45ac64246a52ba36ce0ddcd3c1079ad3e2ff282d3R12) but that is never a good solution and I didn't make enough to fix the problems completelty.
> 
> As I see it there are 5 options:
> 
> 1. Switch back to python-binance, extending it's classes to fix the connection problems
> 2. Switch to our own forked version of python-binance, which fixes the problems and would one day hopefully be merged into the base library
> 3. Do number 2 but wait for the official python-binance library to accept our changes before we merge this PR
> 4. Build our own binance websocket implementation, either in a separate repo or within this one
> 5. Just merge this PR
> 
> I would personally say we do number 5, since this is a major feature that the community is waiting for, and there are a lot of changes that we don't want to have to keep rebasing. Down the road hopefully we can do one of the other options.

Have been testing for more than a week now, ticking like a clock.

**Comment by DmytroLitvinov** [comment]
Great to see it is merged :tada: 
Thank you for your awesome work on it, guys.

@idkravitz , maybe we should also apply some commits from your `master` branch also?

**Comment by israelins85** [comment]
Hello, about that errors on MacOS:

```
python -m binance_trade_bot                                                                           
2021-06-17 23:53:23,891 - crypto_trading_logger - INFO - Starting
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/selector_events.py", line 261, in _add_reader
    key = self._selector.get_key(fd)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/selectors.py", line 193, in get_key
    raise KeyError("{!r} is not registered".format(fileobj)) from None
KeyError: '10 is not registered'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/israelins/Documents/Programacao.nosync/3rd/binance-trade-bot/binance_trade_bot/__main__.py", line 5, in <module>
    main()
  File "/Users/israelins/Documents/Programacao.nosync/3rd/binance-trade-bot/binance_trade_bot/crypto_trading.py", line 18, in main
    manager = BinanceAPIManager(config, db, logger)
  File "/Users/israelins/Documents/Programacao.nosync/3rd/binance-trade-bot/binance_trade_bot/binance_api_manager.py", line 31, in __init__
    self.setup_websockets()
  File "/Users/israelins/Documents/Programacao.nosync/3rd/binance-trade-bot/binance_trade_bot/binance_api_manager.py", line 34, in setup_websockets
    self.stream_manager = BinanceStreamManager(
  File "/Users/israelins/Documents/Programacao.nosync/3rd/binance-trade-bot/binance_trade_bot/binance_stream_manager.py", line 75, in __init__
    self.bw_api_manager.create_stream(
  File "/usr/local/lib/python3.9/site-packages/unicorn_binance_websocket_api/unicorn_binance_websocket_api_manager.py", line 1131, in create_stream
    loop = asyncio.new_event_loop()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/events.py", line 761, in new_event_loop
    return get_event_loop_policy().new_event_loop()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/events.py", line 659, in new_event_loop
    return self._loop_factory()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/unix_events.py", line 54, in __init__
    super().__init__(selector)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/selector_events.py", line 61, in __init__
    self._make_self_pipe()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/selector_events.py", line 112, in _make_self_pipe
    self._add_reader(self._ssock.fileno(), self._read_from_self)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/selector_events.py", line 263, in _add_reader
    self._selector.register(fd, selectors.EVENT_READ,
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/selectors.py", line 523, in register
    self._selector.control([kev], 0, 0)
TypeError: changelist must be an iterable of select.kevent objects
^CException ignored in: <module 'threading' from '/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py'>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 1428, in _shutdown
    lock.acquire()
  File "/usr/local/lib/python3.9/site-packages/engineio/client.py", line 36, in signal_handler
    return original_signal_handler(sig, frame)
  File "/usr/local/lib/python3.9/site-packages/socketio/client.py", line 26, in signal_handler
    return original_signal_handler(sig, frame)
KeyboardInterrupt: 
Exception ignored in: <function BaseEventLoop.__del__ at 0x10d417d30>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 683, in __del__
    self.close()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/unix_events.py", line 63, in close
    if self._signal_handlers:
AttributeError: '_UnixSelectorEventLoop' object has no attribute '_signal_handlers'
```

I see in another thread this is python 3.9 fault, how can I rollback to python 3.8 using brew?

**Comment by oliver-zehentleitner** [comment]
Late maintainer follow-up for anyone reading this thread now: the async/threading/websocket core of UBWA was rewritten about two years ago. To be precise about what changed and what didn't — each stream still runs in its own thread, and that's intentional for a simple reason: an SDK that blocks the caller is a bad SDK. Threading is how you give user code a non-blocking interface to a long-lived I/O lifecycle. Each stream in its own thread also means independent reconnects don't stall each other, and user code doesn't have to live inside asyncio if it doesn't want to — but it can.

What got reworked is the part the comments here actually targeted — structure, readability, and the asyncio integration on the consumer side. The recommended pattern today is the `asyncio_queue` API: `await ubwa.get_stream_data_from_asyncio_queue(stream_id)` with `asyncio_queue_task_done(stream_id)` for consumer-side backpressure, inside a manager that runs as a context manager. On the performance side, UBS now ships as a compiled C extension (Cython) with multi-arch wheels, which is roughly the opposite end of "highly non-optimized."

One thing worth flagging for anyone working on the same class of problem this PR was solving: the original fix here had to listen for connect events manually and re-fetch state via REST to recover from disconnect-induced desync. That pattern is now a first-class API in UBWA — `stream_signals` emit `CONNECT`, `FIRST_RECEIVED_DATA`, `DISCONNECT`, `STOP`, and `STREAM_UNREPAIRABLE` as structured events, so the recovery logic doesn't have to be reverse-engineered from logs anymore.

In hindsight: this PR didn't switch to UBWA because of the architecture, it switched because of the connection-state visibility in the logs. That operational reason was the right one then, and the same need now has a proper API behind it.

— Oliver

**Pull request merged**